Game Rules

geoffthemedio opened the pull request:
Adds game rules manager and UI support.

Rules are stuff that can be configured about how the game works, eg. number of combat rounds, if experimentors are generated, if all techs should cost 50% more than default, etc.

geoffthemedio committed:
grooming

geoffthemedio committed:
-Made OptionsDB::Option public, so it can be used in...
-Added GameRules manager.

geoffthemedio committed:
Used GameRules to store cheap resaerch / production toggles, and number of combat rounds.

geoffthemedio committed:
grooming

geoffthemedio committed:
grooming

geoffthemedio committed:
Added Option::SetToDefault function.

geoffthemedio committed:
Moved adding of combat length rule from CombatSystem.cpp to CombatEvents.cpp so it will be present on clients as well as the server.

geoffthemedio committed:
-Changed storage of game rules in GalaxySetupData to a vector of pairs of strings.
-Setting options by passing in vector of pairs of strings.

geoffthemedio committed:
Renamed, added stringtable entries for game rule names and descriptions.

geoffthemedio committed:
-Added rule description and validator getters.
-Made GameRules constructor private.

geoffthemedio committed:
Replaced some types with auto.

geoffthemedio committed:
Added (nonfunctional) UI widgets to galaxy setup screen to control some game rules.

geoffthemedio committed:
-Connected rule change signals.
-Implemented disable function.

geoffthemedio committed:
Connected rule widget change signals to response function that stores the set rule.

geoffthemedio committed:
Extract set game rules from galaxy setup screen and pass to server so they take effect in-game.

geoffthemedio committed:
Set game rules in client when receiving game start message from server.

geoffthemedio committed:
grooming

geoffthemedio committed:
Added rule for reseeding PRNG.

geoffthemedio committed:
grooming

geoffthemedio committed:
Adding functions to partly automate adding game rule UI controls.

geoffthemedio committed:
Added function to add double spin rule widgets.

LGM-Doyle reviewed (commented):
Great addition.

It works for me.  All techs, buildings and ships can be researched for 1 point and in 1 turn.

geoffthemedio committed:
grooming

geoffthemedio committed:
Added support for querying game rules within ComplexValueRef variants.

geoffthemedio committed:
Added GameRule value ref parser support.

geoffthemedio committed:
Added GameRules::Empty

geoffthemedio committed:
Partly implemented game rules FOCS parser.

geoffthemedio committed:
Added commented out example sitrep FOCS referring to game rules.

geoffthemedio committed:
grooming

geoffthemedio committed:
Added some parser tokens, renamed TrueFalse to Toggle.

geoffthemedio committed:
-more game rule parsing
-test cases

geoffthemedio committed:
-Renamed Enabled / Disabled to On / Off for Toggle rules.
-Tweaked debug output for string lists.
-Fixed, uncommented string rule script.

geoffthemedio committed:
Added scripted rule for hull cost factor to hull costs.

geoffthemedio committed:
Made GameRules::Rule a derived class of OptionsDB::Option in order to add a rule type enum to it, to keep track of what type the rule is.

geoffthemedio committed:
Added functions to check if a rule exists with a particular type, and to get the type of a rule.

geoffthemedio committed:
Partial exposing of game rules to Python API.

geoffthemedio committed:
grooming

geoffthemedio committed:
Exposed GameRules getter to Python API.

geoffthemedio committed:
Added function to check if rule is internal.

geoffthemedio committed:
-Replaced TEST with SCRIPTED rules tab, containing all scripted rules.
-Added test rule strings.

geoffthemedio committed:
-Switched to using std::to_string for displaying Spin values as text.
-Fixed issue with incorrect spin increments when minimum value of range was substantially above 0.
-grooming

geoffthemedio committed:
Implemented CUISpin<T>::SetEditTextFromValue() with T=double specific case using fixed precision.

geoffthemedio committed:
Set increment to 0.1 (were 1.0) in validators for CUISpin<double>.

geoffthemedio committed:
Make labels visible for spinners in rules GUI.

geoffthemedio committed:
-Added droplist of string options GUI widget.
-safety checks

geoffthemedio committed:
Safety try-catch and error logging when trying to set invalid string to an game rule.

geoffthemedio committed:
Adjusted test rule to use valid stringtable keys as allowed / default values.

geoffthemedio committed:
Logging out input rules strings and resulting set rules.

geoffthemedio committed:
Fixed / implemented setting up and using specialized rows for rule string options.

geoffthemedio committed:
-Removed STRING_LIST rule type, as it can probably be more simply implemented as a series of toggles.
-Moved enum wrapping into EnumWrapper.cpp

geoffthemedio committed:
Test / Debug logging of defined game rules in AI script.

geoffthemedio committed:
Added category to game rules, used to populate pages in the GUI.

geoffthemedio committed:
-Removed some test rules.
-Implemented part cost scale rule.

geoffthemedio committed:
Removed duplicate string.

geoffthemedio committed:
Added tooltips to rule widget label text.

geoffthemedio committed:
Replaced auto with explicit iterator types to hopefully fix XCode build error.

LGM-Doyle reviewed (commented):
Thanks for making the previous changes to use `unordered_map`.

As usual the code looks good.  It compiles, changes production, research times and duration of combat as expected on linux.

I've added some questions about the usage of `engine_internal` and the enumeration of the rule types.

geoffthemedio committed:
Removed previously-commented-out line.

Vezzra committed:
Update Xcode project

Vezzra reviewed (approved):
> Further comments?

I've pushed an update of the Xcode project, builds and runs fine with that on OSX.

As far as I'm concerned, LGTM.
